### ISSUE TITLE
logsumexp with sign indicator - enable calculation with negative signs

### ISSUE THREAD

**Issue by ingmarschuster** [numpy/numpy#5652]
So my problem has been the following: I had to calculate the logsumexp of some numbers in logspace for accuracy reasons, but some of the elements of the array have negative sign. I implemented the following code, which you might want to integrate into the upstream logsomexp implementation of numpy.
The "sign" argument is a matrix assumed to be filled with +1 or -1 depending on the sign of the number before the log transformation. 
```
from numpy import exp, log, sqrt
from scipy.misc import logsumexp
import numpy as np
import scipy.stats as stats


def log_sign(a):
    a = np.array(a)
    sign_indicator = ((a < 0 ) * -2 + 1)
    return (log(np.abs(a)), sign_indicator)

def exp_sign(a, sign_indicator):
    return exp(a) * sign_indicator    
    
def logsumexp_sign(a, axis = None, b = None, sign = None):
    if sign is None:
        abs_res = logsumexp(a, axis = axis, b = b)
        sign_res = np.ones_like(abs_res)
        return (abs_res, sign_res)
    else:
        if not (np.abs(sign) == 1).all():
            raise ValueError("sign arguments expected to contain only +1 or -1 elements")
        
        m = np.copy(a)
        m[sign == -1] = -np.infty # don't look at negative numbers
        m = logsumexp(m, axis = axis, b = b)
        m[np.isnan(m)] = -np.infty # replace NaNs resulting from summing infty
        
        s = np.copy(a)
        s[sign == +1] = -np.infty # don't look at positive numbers
        s = logsumexp(s, axis = axis, b = b)
        s[np.isnan(s)] = -np.infty # replace NaNs resulting from summing infty
        

        sign_res =  np.ones(np.broadcast(m, s).shape) #default to positive sign
        abs_res  = -np.infty * sign_res #default to log(0)
        
        idx = np.where(m > s)
        abs_res[idx] = log(1 - exp(s[idx] - m[idx])) + m[idx]
        
        idx = np.where(m < s)
        sign_res[idx] = -1
        abs_res[idx] = log(1 - exp(m[idx] - s[idx])) + s[idx]
        return (abs_res, sign_res)
```
I'm not sure I passed on the `b` parameter correctly (I never made use of it before). Here is a test method:
```
def test_logsumexp_sign():    
    for ax in (0, 1):
        for a in [np.array([[1,1,-1], [-1,-1,1]]),
                  np.array([[1,1,1], [-1,-1,-1]])]:
            (la, sa) = log_sign(a)
            log_res = logsumexp_sign(la, axis = ax, sign = sa)
            res = exp_sign(*log_res)
            if not np.all(np.abs(res - np.sum(a, axis = ax)).mean() <= 0.1):
                raise RuntimeError("problem when summing accros", ax,
                                   "for", repr(a), "getting", log_res)
```

**Comment by njsmith** [comment]
I'm actually surprised this is producing any difference in accuracy:
logsumexp makes some modest attempts at reducing rounding error but I
didn't think it would be that much better than plain sum. Can you give any
example where this produces more accurate results than plain sum?

(Note that Python's standard math.fsum function provides exactly accurate
sums. It may be too slow for you to use directly, but it can be useful when
trying to figure out how much error some other method produced. Note also
that modern numpy.sum's accuracy can depend on the memory layout of your
array; specifically, when summing slices of a 2d array, summing along
columns might get higher accuracy than summing along rows, or vice-versa.
Simple 1d arrays should always get the higher accuracy.)

**Comment by ingmarschuster** [comment]
Yes, so basically when computing with very small numbers (as for example in Monte Carlo sampling from a distribution), it is very common to keep the numbers in logspace. Most of the time, you only want to add up numbers in an array which could be done using `log(sum(exp(a)))` - this however has accuracy problems, which the (already existing) `scipy.misc.logsumexp` solves. Now I had to compute the when estimating a very small number, `estimate - true_number`, which could become negative (and very small). To be able to keep the numbers in logspace, I came up  with the suggested implementation `logsumexp_sign` - which could actually be integrated with  `scipy.misc.logsumexp`, and I think thats the preferred solution.

**Comment by argriffing** [comment]
For what it's worth, there is an [`slogdet`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.slogdet.html), so maybe `slogsumexp`?

**Comment by njsmith** [comment]
Right, I understand all that. My question is whether you've double-checked
that all this actually works as well as you think it should :-). If you
look under the hood then you'll see that logsumexp is not actually doing
anything very fancy to reduce rounding error. It can help in some cases but
isn't magical.

**Comment by ingmarschuster** [comment]
Yes, I mean, of course `logsumexp` works way better for my practical work, or else I wouldn't use it. The example I gave is not contrived in terms of values that `-sp.stats.gamma.rvs(10,100000, size=100)` produces - I actually get values in that range in day to day problems.

And `logsumexp_sign` was an extension I needed that I thought might make sense for other people as well.

**Comment by njsmith** [comment]
Odd, your whole example just doesn't exist in the version of your message
that github emailed me -- I had no idea it existed until I went to the
website just now in confusion. (Did you edit it in after posting or
anything like that?) I'll take a look...

**Comment by juliantaylor** [comment]
somewhat related gh-4260, and one my many unfinished branches for it https://github.com/juliantaylor/numpy/tree/logaddexp
adding sign would be a bit problematic as you can't add more arguments to ufuncs. Though as logsumexp is a scipy function already maybe it makes more sense to have the sign variant there too

**Comment by njsmith** [comment]
I wonder if a better api for this would be to add logsubexp analogous to
logaddexp, so one could write

logsubexp(logsumexp(arr[sign == 1], logsumexp(arr[sign == -1]))

Afaict this would be equivalent to the original code, assuming logsubexp
has overflow protection similar to that in logsumexp. It would have to
return a (value, sign) tuple of course, similar to slogdet.

As a procedural matter, you might want to file a bug with scipy, since
they're the ones who provide the logsumexp interface and would be the ones
you'd need to convince to add your original api proposal. Numpy also
provides a logsumexp (we call it logaddexp.reduce), but it doesn't have
special overflow protection (that's gh-4260), and it since it's a lower
level tool it doesn't really have any good way to add an extra argument to
implement your proposed api. OTOH numpy could implement other APIs, like
the one I suggested above.

In the far off future I wonder if the right solution isn't to make a
log-space dtype (as a third party library presumably), with signed and
unsigned variants, where + is implemented as logaddexp, * is implemented as
+, etc.0

**Comment by njsmith** [comment]
I'm sorry, I have told a lie. Numpy's logaddexp.reduce does do
underflow/overflow avoidance, and has since 2009. So wanting to solve this
using pure numpy would be totally reasonable, but the
logsumexp-with-extra-args approach is still probably not the ideal numpy
api.

**Comment by ingmarschuster** [comment]
Actually, I was thinking that a logspace type would make a lot of sense, just like you. You are right that it probably makes sense to file a bug with scipy though. I actually had a `logsubexp` implemented though, so I'm posting it here. It also uses sign arguments and is correct; however it might be a bit slow, since it is not ideal wrt vectorization.
```
def logsubtrexp(minuend, subtrahend, sign_minuend = None, sign_subtrahend = None):

    if sign_minuend is None:
        sign_minuend = np.ones(minuend.shape)
    if sign_subtrahend is None:
        sign_subtrahend = np.ones(subtrahend.shape)
    if not (minuend.shape == sign_minuend.shape and subtrahend.shape == sign_subtrahend.shape):
        raise ValueError("sign arguments expected be of same shape as corresponding log-matrices")
    if not (np.abs(sign_minuend).all() and np.abs(sign_subtrahend).all()):
        raise ValueError("sign arguments expected to contain only +1 or -1 elements")
        
    b = np.broadcast(minuend, subtrahend)
    s_b = np.broadcast(sign_minuend, sign_subtrahend)
    abs_res = np.empty(b.shape)
    sign_res = np.empty(b.shape)
    
    for i in range(b.size):
        (m, s) = b.next()
        (sign_m, sign_s) = s_b.next()
        if sign_m > sign_s: # sign_m == 1 and sign_s == -1
            # this is equivalent to logsumexp(m, s)
            #print("sign_m > sign_s")
            sign_res.flat[i] = 1
            abs_res.flat[i] = logsumexp((m,s))
        elif sign_m < sign_s: # sign_m == -1 and sign_s == 1
            #print("sign_m < sign_s")
            sign_res.flat[i] = -1
            abs_res.flat[i] = logsumexp((m,s))
        else:
            #signs are eqal
            if m == s:                
                sign_res.flat[i] = 1
                abs_res.flat[i] = log(0)
            else:
                if sign_m == -1:
                    if m > s:
                        #print("m >= s")
                        sign_res.flat[i] = -1
                        abs_res.flat[i] = log(1 - exp(s - m)) + m
                    elif m < s:
                        #print("m < s")
                        sign_res.flat[i] = 1
                        abs_res.flat[i] = log(1 - exp(m - s)) + s
                else:# sign_m == 1
                    if m > s:
                        #print("m >= s")
                        sign_res.flat[i] = 1
                        abs_res.flat[i] = log(1 - exp(s - m)) + m
                    elif m < s:
                        #print("m < s")
                        sign_res.flat[i] = -1
                        abs_res.flat[i] = log(1 - exp(m - s)) + s
        #print(sign_m*exp(m),  sign_s*exp(s),  sign_m*exp(m) - sign_s*exp(s), sign_res.flat[i] * exp(abs_res.flat[i]))
    
    return (abs_res, sign_res)
```

And for testing (using the `log_sign` and `exp_sign` from above):
```
def test_log_subtr_add_exp():    
    for (m, s) in [(np.array([[-10.,   3.],[ -1.,   5.]]), np.array([-5.5,  4. ])),
                   (np.array([[10.,   -3.],[ 1.,   -5.]]), np.array([5.5,  -4. ])),
                   (3, 10),
                   (3, 0),
                   (-30, 0), 
                   (np.arange(3).astype(float), 1),
                   ((np.arange(12).reshape((2,2,3)) - 4).astype(float),1)]:
        (lm, sm) = log_sign(m)
        (ls, ss) = log_sign(s)
        (subtr, subtr_s) = logsubtrexp(lm, ls, sm, ss)
        assert((np.abs((m - s) - exp_sign(subtr, subtr_s)) < 1e-14).all())
```

**Comment by mdhaber** [comment]
I've needed this sort of thing in `scipy.stats` many times, but I've found the `b` (weights) and `return_sign` interface of `logsumexp` to be cumbersome. Instead, I add `pi*1j` to the elements that are to be subtracted and just use complex math. Yes, it's a little slower, but I think the elegance/simplicity is worth it. 

It seems like the interface question is what slowed down this issue. To close this, can we just enable support for complex input to `logaddexp`? If I'm not mistaken, little else about the implementation would need to change.
